File: modules/local/soupx/templates/soupx.py
#!/usr/bin/env python3

import scanpy
import anndata2ri
import rpy2
import rpy2.robjects as ro
import platform
soupx = ro.packages.importr('soupx')

def format_yaml_like(data: dict, indent: int = 0) -> str:
    """Formats a dictionary to a YAML-like string.

    Args:
        data (dict): The dictionary to format.
        indent (int): The current indentation level.

    Returns:
        str: A string formatted as YAML.
    """
    yaml_str = ""
    for key, value in data.items():
        spaces = "  " * indent
        if isinstance(value, dict):
            yaml_str += f"{spaces}{key}:\\n{format_yaml_like(value, indent + 1)}"
        else:
            yaml_str += f"{spaces}{key}: {value}\\n"
    return yaml_str

adata = scanpy.read_h5ad("${h5ad}")

adata_pp = adata.copy()
scanpy.pp.normalize_per_cell(adata_pp)
scanpy.pp.log1p(adata_pp)

scanpy.pp.pca(adata_pp)
scanpy.pp.neighbors(adata_pp)
scanpy.tl.leiden(adata_pp, key_added="soupx_groups")

soupx_groups = adata_pp.obs["soupx_groups"]
del adata_pp

cells = adata.obs_names
genes = adata.var_names
data = adata.X.T

adata_raw = scanpy.read_h5ad("${raw}")
data_raw = adata_raw.X.T
del adata_raw

r_data = anndata2ri.py2rpy(data)
r_data_raw = anndata2ri.py2rpy(data_raw)
ro.r("function(data, genes, cells) { rownames(data) <- genes; colnames(data) <- cells }")(data, genes, cells)

sc = soupx.SoupChannel(r_data_raw, r_data, calcSoupProfile = False)

soupProf = ro.r("function(data) { data.frame(row.names = rownames(data), est = rowSums(data)/sum(data), counts = rowSums(data)) }")(r_data)
sc = soupx.setSoupProfile(sc, soupProf)
sc = soupx.setClusters(sc, soupx_groups)
sc = soupx.autoEstCont(sc, doPlot = False)
out = soupx.adjustCounts(sc, roundToInt = True)

adata.layers["ambient"] = out.T

adata.write_h5ad("${prefix}.h5ad")

# Versions

versions = {
    "${task.process}": {
        "python": platform.python_version(),
        "scanpy": scanpy.__version__,
        "anndata2ri": anndata2ri.__version__,
        "rpy2": rpy2.__version__,
        "soupx": soupx.__version__,
    }
}

with open("versions.yml", "w") as f:
    f.write(format_yaml_like(versions))

File: modules/local/soupx/templates/soupx.R
#!/usr/bin/env Rscript

library(SoupX)
library(anndataR)
library(Seurat)

# Read the AnnData objects and convert directly to Seurat
adata <- read_h5ad("${h5ad}")
seu <- adata\$as_Seurat(x_mapping = "counts")

# Read the raw AnnData object
adata_raw <- read_h5ad("${raw}")
seu_raw <- adata_raw\$as_Seurat(x_mapping = "counts")

# Get layer to use (default to "X" if not specified)
use_layer <- "${input_layer}"
if (use_layer != "X") {
    if (use_layer %in% names(adata\$layers)) {
        # Replace counts in the Seurat object with the specified layer
        counts_matrix <- t(adata\$layers[[use_layer]])
        rownames(counts_matrix) <- rownames(seu)
        colnames(counts_matrix) <- colnames(seu)
        seu[["RNA"]] <- CreateAssayObject(counts = counts_matrix)
    } else {
        stop(paste("Layer", use_layer, "not found in AnnData object"))
    }
}

# Preprocessing with Seurat workflow
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu)
seu <- ScaleData(seu)
seu <- RunPCA(seu)
seu <- FindNeighbors(seu)
seu <- FindClusters(seu, resolution = ${cluster_resolution})
soupx_groups <- Idents(seu)

# Create SoupChannel
data <- GetAssayData(seu, layer = "counts")
data_raw <- GetAssayData(seu_raw, layer = "counts")
sc <- SoupChannel(data_raw, data, calcSoupProfile = FALSE)

# Set soup profile - following SoupX vignette approach
soupProf <- data.frame(
    row.names = rownames(data),
    est = rowSums(data)/sum(data),
    counts = rowSums(data)
)
sc <- setSoupProfile(sc, soupProf)

# Set clusters and estimate contamination
sc <- setClusters(sc, soupx_groups)
sc <- autoEstCont(sc, doPlot = FALSE)

# Adjust counts
out <- adjustCounts(sc, roundToInt = FALSE)

# Update the original AnnData object with ambient layer
if ("${output_layer}" == "X") {
    adata\$X <- t(out)
} else {
    adata\$layers[["${output_layer}"]] <- t(out)
}

# Save the output
write_h5ad(adata, "${prefix}.h5ad")

# Write version information
writeLines(
    c(
        '"${task.process}":',
        paste('    r:', paste(R.Version()\$major, R.Version()\$minor, sep = ".")),
        paste('    anndataR:', as.character(packageVersion('anndataR'))),
        paste('    SoupX:', as.character(packageVersion('SoupX'))),
        paste('    Seurat:', as.character(packageVersion('Seurat')))
    ),
    'versions.yml'
)
